stopes/pipelines/asr_bleu/transcribe_audio.py

```diff
@@ -0,0 +1,213 @@
+            sample_rate: the target sampling rate for the ASR model
+            normalize_input: the bool value for the ASR model
+
+        Returns:
+            audio_waveform: the audio waveform as a torch.Tensor object
+        """
+        audio_waveform, sampling_rate = torchaudio.load(audio_path)
+        if audio_waveform.dim == 2:
+            audio_waveform = audio_waveform.mean(-1)
+        if sample_rate != sampling_rate:
+            audio_waveform = torchaudio.functional.resample(audio_waveform, sampling_rate, sample_rate)
+        if normalize_input:
+            # following fairseq raw audio dataset
+            audio_waveform = torch.nn.functional.layer_norm(audio_waveform, audio_waveform.shape)
+
+        return audio_waveform
+
+    @torch.inference_mode()
+    def _compute_emissions(self, audio_input: torch.Tensor, asr_model: fairseq.models.FairseqEncoder, cuda) -> torch.Tensor:
+        """
+        Compute the emissions for either fairseq or huggingface asr model
+
+        Args:
+            audio_input: the input audio waveform
+            asr_model: the ASR model to use for computing emissions
+
+        Returns:
+            emissions: the logits of the encoded prediction.
+        """
+
+        if cuda:
+            audio_input = audio_input.to("cuda")
+        if isinstance(asr_model, fairseq.models.wav2vec.wav2vec2_asr.Wav2VecCtc):
+            padding_mask = lengths_to_padding_mask(torch.tensor([audio_input.numel()]))
+            emissions = asr_model.w2v_encoder(audio_input, padding_mask)["encoder_out"].transpose(0, 1)
+        else:
+            emissions = asr_model(audio_input).logits
+
+        return emissions
+
+
+    def _decode_emissions(self,emissions: torch.Tensor, decoder, post_process_fn) -> str:
+        """
+        Decode the emissions and apply post-process functions
+
+        Args:
+            emissions: the input Tensor object
+            decoder: ASR decoder
+            post_process_fn: post-process function for the ASR output
+
+        Returns:
+            hypo: the str as the decoded transcriptions
+        """
+
+        emissions = emissions.cpu()
+        results = decoder(emissions)
+
+        # assuming the lexicon-free decoder and working with tokens
+        hypo = decoder.idxs_to_tokens(results[0][0].tokens)
+        hypo = post_process_fn(hypo)
+
+        return hypo
+
+    def _merge_tailo_init_final(self, text):
+        """
+        Hokkien ASR hypothesis post-processing.
+        """
+        sps = text.strip().split()
+        results = []
+        last_syllable = ""
+        for sp in sps:
+            if sp == "NULLINIT" or sp == "nullinit":
+                continue
+            last_syllable += sp
+            if sp[-1].isnumeric():
+                results.append(last_syllable)
+                last_syllable = ""
+        if last_syllable != "":
+            results.append(last_syllable)
+        return " ".join(results)
+
+    def _remove_tone(self, text):
```

Comment:
Is this used anywhere?

Reply:
No, it was left over from the old codebase where it was also not used. I'll remove it for now.

Reply:
Resolved with https://github.com/facebookresearch/stopes/pull/52/commits/c422fc7985fe8c84f42b7829d3c2efaadf3eef9a